pki/util.go

```diff
@@ -0,0 +1,208 @@
+	ips := []net.IP{}
+	dnsNames := []string{}
+	for _, host := range cpHosts {
+		// Check if node address is a valid IP
+		if nodeIP := net.ParseIP(host.Address); nodeIP != nil {
+			ips = append(ips, nodeIP)
+		} else {
+			dnsNames = append(dnsNames, host.Address)
+		}
+
+		// Check if node internal address is a valid IP
+		if len(host.InternalAddress) != 0 && host.InternalAddress != host.Address {
+			if internalIP := net.ParseIP(host.InternalAddress); internalIP != nil {
+				ips = append(ips, internalIP)
+			} else {
+				dnsNames = append(dnsNames, host.InternalAddress)
+			}
+		}
+		// Add hostname to the ALT dns names
+		if len(host.HostnameOverride) != 0 && host.HostnameOverride != host.Address {
+			dnsNames = append(dnsNames, host.HostnameOverride)
+		}
+	}
+	ips = append(ips, net.ParseIP("127.0.0.1"))
+	ips = append(ips, KubernetesServiceIP)
+	dnsNames = append(dnsNames, []string{
+		"localhost",
+		"kubernetes",
+		"kubernetes.default",
+		"kubernetes.default.svc",
+		"kubernetes.default.svc." + clusterDomain,
+	}...)
+	return &cert.AltNames{
+		IPs:      ips,
+		DNSNames: dnsNames,
+	}
+}
+
+func (c *CertificatePKI) ToEnv() []string {
+	env := []string{
+		c.CertToEnv(),
+		c.KeyToEnv(),
+	}
+	if c.Config != "" {
+		env = append(env, c.ConfigToEnv())
+	}
+	return env
+}
+
+func (c *CertificatePKI) CertToEnv() string {
+	encodedCrt := cert.EncodeCertPEM(c.Certificate)
+	return fmt.Sprintf("%s=%s", c.EnvName, string(encodedCrt))
+}
+
+func (c *CertificatePKI) KeyToEnv() string {
+	encodedKey := cert.EncodePrivateKeyPEM(c.Key)
+	return fmt.Sprintf("%s=%s", c.KeyEnvName, string(encodedKey))
+}
+
+func (c *CertificatePKI) ConfigToEnv() string {
+	return fmt.Sprintf("%s=%s", c.ConfigEnvName, c.Config)
+}
+
+func getEnvFromName(name string) string {
+	return strings.Replace(strings.ToUpper(name), "-", "_", -1)
+}
+
+func getKeyEnvFromEnv(env string) string {
+	return fmt.Sprintf("%s_KEY", env)
+}
+
+func getConfigEnvFromEnv(env string) string {
+	return fmt.Sprintf("KUBECFG_%s", env)
+}
+
+func GetEtcdCrtName(address string) string {
+	newAddress := strings.Replace(address, ".", "-", -1)
+	return fmt.Sprintf("%s-%s", EtcdCertName, newAddress)
+}
+
+func GetCertPath(name string) string {
+	return fmt.Sprintf("%s%s.pem", CertPathPrefix, name)
+}
+
+func GetKeyPath(name string) string {
+	return fmt.Sprintf("%s%s-key.pem", CertPathPrefix, name)
+}
+
+func GetConfigPath(name string) string {
+	return fmt.Sprintf("%skubecfg-%s.yaml", CertPathPrefix, name)
+}
+
+func GetCertTempPath(name string) string {
+	return fmt.Sprintf("%s%s.pem", TempCertPath, name)
+}
+
+func GetKeyTempPath(name string) string {
+	return fmt.Sprintf("%s%s-key.pem", TempCertPath, name)
+}
+
+func GetConfigTempPath(name string) string {
+	return fmt.Sprintf("%skubecfg-%s.yaml", TempCertPath, name)
+}
+
+func ToCertObject(componentName, commonName, ouName string, cert *x509.Certificate, key *rsa.PrivateKey) CertificatePKI {
```

Comment:
is this function ever called from other packages? if not, consider reducing the visibility.

Reply:
Yeah unfortunately I call ToCertObject from cluster package when i am regenerating kubeAPI certificate, we do that because when adding new kubeapi node, we need to create a new certificate for it with its IP (internal and public) in the ALTNames of the certificate